rust-symcrypt/src/ecdh.rs

```diff
@@ -0,0 +1,257 @@
+                self.key.inner,
+                symcrypt_sys::_SYMCRYPT_ECPOINT_FORMAT_SYMCRYPT_ECPOINT_FORMAT_XY,
+            );
+
+            let mut pub_key_bytes = vec![0u8; pub_key_len as usize];
+
+            match symcrypt_sys::SymCryptEckeyGetValue(
+                self.key.inner,
+                std::ptr::null_mut(),
+                0 as symcrypt_sys::SIZE_T,
+                pub_key_bytes.as_mut_ptr(),
+                pub_key_len as symcrypt_sys::SIZE_T,
+                num_format,
+                ec_point_format,
+                0, // No flags allowed
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(pub_key_bytes),
+                err => Err(err.into()),
+            }
+        }
+    }
+
+    pub fn ecdh_secret_agreement(
+        private: &EcDh,
+        public: &EcDh,
+    ) -> Result<EcDhSecretAgreement, SymCryptError> {
+        let num_format = get_num_format(private.curve_type);
+
+        unsafe {
+            // SAFETY: FFI calls
+            let secret_length = symcrypt_sys::SymCryptEcurveSizeofFieldElement(private.key.curve.0);
+            let mut secret = vec![0u8; secret_length as usize];
+
+            match symcrypt_sys::SymCryptEcDhSecretAgreement(
+                private.key.inner,
+                public.key.inner,
+                num_format,
+                0,
+                secret.as_mut_ptr(),
+                secret.len() as symcrypt_sys::SIZE_T,
+            ) {
+                symcrypt_sys::SYMCRYPT_ERROR_SYMCRYPT_NO_ERROR => Ok(EcDhSecretAgreement(secret)),
+                err => Err(err.into()),
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::symcrypt_init;
+
+    /// symcrypt_sys::SymCryptModuleInit() must be called via lib.rs in order to initialize the callbacks for
+    /// SymCryptEcurveAllocate, SymCryptEckeyAllocate, SymCryptCallbackAlloc, etc.
+    #[test]
+    fn test_ecdh_nist_p256() {
+        symcrypt_init(); // must run symcrypt_init for the alloc callbacks to initialize
+        let mut ecdh_1_private = EcDh::new(CurveType::NistP256).unwrap();
+        let mut ecdh_2_private = EcDh::new(CurveType::NistP256).unwrap();
+
+        let public_bytes_1 = ecdh_1_private.get_public_key_bytes().unwrap();
+        let public_bytes_2 = ecdh_2_private.get_public_key_bytes().unwrap();
+
+        let ecdh_1_public =
+            EcDh::from_public_key_bytes(CurveType::NistP256, &public_bytes_1.as_slice()).unwrap();
+        let ecdh_2_public =
+            EcDh::from_public_key_bytes(CurveType::NistP256, &public_bytes_2.as_slice()).unwrap();
+
+        let secret_agreement_1 =
+            EcDh::ecdh_secret_agreement(&ecdh_1_private, &ecdh_2_public).unwrap();
+        let secret_agreement_2 =
+            EcDh::ecdh_secret_agreement(&ecdh_2_private, &ecdh_1_public).unwrap();
+
+        assert_eq!(secret_agreement_1.0, secret_agreement_2.0);
```

Comment:
>assert_eq!(secret_agreement_1.0, secret_agreement_2.0); [](http://example.com/codeflow?start=8&length=55)

Can `EcDhSecretAgreement` objects not be compared directly? They're basically just wrappers/typedefs/aliases (not sure the correct Rust terminology) for vectors, so it seems like that should work...

Reply:
You have to opt in to being able to compare them. A lot of types this kind of newtype wrapper is done specifically to remove the ability to compare by equality. But if you add `#[derive(Eq)]` (and a few related traits) to `EcDhSecretAgreement` then you'll get the ability back to compare them for equality.